frame/dex/src/lib.rs

```diff
@@ -524,14 +507,59 @@ pub mod pallet {
 			to: &T::AccountId,
 			amount: AssetBalanceOf<T>,
 			keep_alive: bool,
-		) -> Result<<T as pallet::Config>::AssetBalance, DispatchError> {
+		) -> Result<T::AssetBalance, DispatchError> {
+			Self::deposit_event(Event::Transfer {
+				from: from.clone(),
+				to: to.clone(),
+				asset: asset_id,
+				amount,
+			});
 			match asset_id {
 				MultiAssetId::Native => T::Currency::transfer(from, to, amount, keep_alive),
 				MultiAssetId::Asset(asset_id) =>
 					T::Assets::transfer(asset_id, from, to, amount, keep_alive),
 			}
 		}
 
+		fn do_swap(
+			sender: &T::AccountId,
+			amounts: &Vec<AssetBalanceOf<T>>,
+			path: &BoundedVec<MultiAssetId<T::AssetId>, T::MaxSwapPathLength>,
+			send_to: &T::AccountId,
+			keep_alive: bool,
+		) -> Result<(), DispatchError> {
+			if let Some(&[asset1, asset2]) = path.get(0..2) {
+				let pool_id = Self::get_pool_id(asset1, asset2);
+				let pool_account = Self::get_pool_account(pool_id);
+				let first_amount = amounts.first().expect("Always has more than one element");
+
+				Self::transfer(asset1, sender, &pool_account, *first_amount, keep_alive)?;
+
+				let path_len = path.len() as u32;
```

Comment:
I think this is a perfect place to use slice's windows function:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=73985c0dfc8572a27b1e220776152192

Reply:
tempted to unroll the last round of the loop for readability. Might be less ifs that way - a bit clearer.